subsys/net/lib/lwm2m/ipso_push_button.c

```diff
@@ -166,6 +235,223 @@ static struct lwm2m_engine_obj_inst *button_create(uint16_t obj_inst_id)
+{
+	switch (pin_pos) {
+	case BIT(SW0_GPIO_PIN): return 0;
+#if DT_NODE_HAS_STATUS(SW1_NODE, okay)
+	case BIT(SW1_GPIO_PIN): return 1;
+#endif
+#if DT_NODE_HAS_STATUS(SW2_NODE, okay)
+	case BIT(SW2_GPIO_PIN): return 2;
+#endif
+#if DT_NODE_HAS_STATUS(SW3_NODE, okay)
+	case BIT(SW3_GPIO_PIN): return 3;
+#endif
+	}
+
+	LOG_ERR("No match for GPIO pin 0x%08x\n", pin_pos);
+
+	return -EINVAL;
+}
+
+static void button_off_timer(struct k_timer *work)
+{
+	struct sw *sw = CONTAINER_OF(work, struct sw, button_timer);
+	char pathstr[15];
+	uint32_t val = 0U;
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(inst); i++) {
+		if (!sw->swi[i].pressed)
+			continue;
+
+		val = gpio_pin_get(sw->gpio, sw->swi[i].pin);
+		if (val < 0) {
+			LOG_ERR("get_pin_value, %d", val);
+			k_timer_stop(&sw->button_timer);
+			continue;
+		}
+		LOG_DBG("switch: %d, val: %d", i, val);
+
+		/* check if button is stil pressed */
+		if (val == 0) {
+			/* restart the timer to sample pin again */
+			k_timer_start(&sw->button_timer, K_SECONDS(1),
+					K_NO_WAIT);
+			continue;
+		}
+
+		sw->swi[i].pressed = false;
+
+		LOG_DBG("switch: %d, pressed: false", i);
+		snprintk(pathstr, sizeof(pathstr), "/%u/%u/%u",
+				IPSO_OBJECT_PUSH_BUTTON_ID, i,
+				BUTTON_DIGITAL_STATE_ID);
+		lwm2m_engine_set_bool(pathstr, !val);
+	}
+}
+
+static void button_pressed_worker(struct k_work *work)
+{
+	struct sw *sw = CONTAINER_OF(work, struct sw, button_work);
+	char pathstr[15];
+	uint32_t val = 0U;
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(inst); i++) {
+		if (!sw->swi[i].pressed)
+			continue;
+
+		val = gpio_pin_get(sw->gpio, sw->swi[i].pin);
+		if (val < 0) {
+			LOG_ERR("get_pin_value, %d", val);
+			k_timer_stop(&sw->button_timer);
+			sw->swi[i].pressed = false;
+			continue;
+		}
+
+		LOG_DBG("switch: %d, val: %d, pressed: true", i, val);
+		snprintk(pathstr, sizeof(pathstr), "/%u/%u/%u",
+				IPSO_OBJECT_PUSH_BUTTON_ID, i,
+				BUTTON_DIGITAL_STATE_ID);
+		lwm2m_engine_set_bool(pathstr, !val);
+
+		k_timer_stop(&sw->button_timer);
+		k_timer_start(&sw->button_timer, K_MSEC(500), K_NO_WAIT);
+	}
+}
+
+static void button_pressed(const struct device *dev, struct gpio_callback *cb,
+		uint32_t pin_pos)
+{
+	struct sw *sw = CONTAINER_OF(cb, struct sw, button_cb);
+	int sw_num = pin_pos_to_sw_num(pin_pos);
+	int res;
+
+	LOG_DBG("pin_pos: 0x%08X, sw_num: %d", pin_pos, sw_num);
+
+	if (sw_num < 0) {
+		return;
+	}
+
+	if (sw_num >= ARRAY_SIZE(inst)) {
+		LOG_INF("No available instaces for sw_num %d", sw_num);
+		return;
+	}
+
+	time = k_uptime_get_32();
+
+	res = pin_pos_to_pin(pin_pos);
+	if (res < 0) {
+		return;
+	}
+	sw->swi[sw_num].pin = (uint8_t)res;
+	sw->swi[sw_num].pressed = true;
+
+	/* debounce the switch */
+	if (time < sw->swi[sw_num].last_time + BUTTON_DEBOUNCE_DELAY_MS) {
+		sw->swi[sw_num].last_time = time;
+		LOG_DBG("switch: %d, debounce", sw_num);
+		return;
+	}
+
+	k_work_submit(&sw->button_work);
+	sw->swi[sw_num].last_time = time;
+}
+
+static void configure_buttons(void)
```

Comment:
I really don't think that button handling including GPIO configuration, interrupt processing and debouncing should be a part of the LwM2M library. I think that users should have liberty to choose which button reports to the server and what is the actual mapping between buttons on the DK and LwM2M object instance.

I'd rather see this code as a part of the LwM2M client sample. The button handling module should make use of the LwM2M APIs to update its status in the server, not the other way.

Reply:
> I'd rather see this code as a part of the LwM2M client sample.

I agree. This kind of functionality should be put to the sample app.

Reply:
You're both right, I agree with you. I'll try to re-implement such functionality but as part of example not as part of library.